src/sys/redox/selector.rs

```diff
@@ -0,0 +1,142 @@
+
+        // offset by 1 to avoid choosing 0 as the id of a selector
+        let id = NEXT_ID.fetch_add(1, Ordering::Relaxed) + 1;
+
+        Ok(Selector {
+            id: id,
+            efd: efd,
+        })
+    }
+
+    pub fn id(&self) -> usize {
+        self.id
+    }
+
+    /// Wait for events from the OS
+    pub fn select(&self, evts: &mut Events, awakener: Token, _timeout: Option<Duration>) -> io::Result<bool> {
+        use std::slice;
+
+        let dst = unsafe {
+            slice::from_raw_parts_mut(
+                evts.events.as_mut_ptr() as *mut u8,
+                evts.events.capacity() * mem::size_of::<syscall::Event>())
+        };
+
+        let cnt = try!(read(self.efd, dst).map_err(super::from_syscall_error))
+                                          / mem::size_of::<syscall::Event>();
+
+        unsafe { evts.events.set_len(cnt); }
+
+        for i in 0..cnt {
+            if evts.get(i).map(|e| e.token()) == Some(awakener) {
+                evts.events.remove(i);
+                return Ok(true);
+            }
+        }
+
+        Ok(false)
+    }
+
+    /// Register event interests for the given IO handle with the OS
+    pub fn register(&self, fd: RawFd, _token: Token, interests: Ready, opts: PollOpt) -> io::Result<()> {
+        let flags = ioevent_to_fevent(interests, opts);
+        fevent(fd, flags).map_err(super::from_syscall_error).map(|_| ())
+    }
+
+    /// Register event interests for the given IO handle with the OS
+    pub fn reregister(&self, fd: RawFd, _token: Token, interests: Ready, opts: PollOpt) -> io::Result<()> {
+        let flags = ioevent_to_fevent(interests, opts);
+        fevent(fd, flags).map_err(super::from_syscall_error).map(|_| ())
+    }
+
+    /// Deregister event interests for the given IO handle with the OS
+    pub fn deregister(&self, fd: RawFd) -> io::Result<()> {
+        fevent(fd, 0).map_err(super::from_syscall_error).map(|_| ())
+    }
+}
+
+fn ioevent_to_fevent(interest: Ready, _opts: PollOpt) -> usize {
+    let mut flags = 0;
+
+    if interest.is_readable() {
+        flags |= EVENT_READ;
+    }
+
+    flags
+}
+
+impl Drop for Selector {
+    fn drop(&mut self) {
+        let _ = close(self.efd);
+    }
+}
+
+pub struct Events {
+    events: Vec<syscall::Event>,
+}
+
+impl Events {
+    pub fn with_capacity(u: usize) -> Events {
+        Events {
+            events: Vec::with_capacity(u)
+        }
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.events.len()
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.events.is_empty()
+    }
+
+    #[inline]
+    pub fn get(&self, idx: usize) -> Option<Event> {
+        self.events.get(idx).map(|event| {
+            let flags = event.flags;
+            let mut kind = Ready::none();
+
+            if flags & EVENT_READ == EVENT_READ {
+                kind = kind | Ready::readable();
+            }
+
+            let token = self.events[idx].id;
```

Comment:
**Warning: The token still needs to be correctly implemented similarly to `udata` in `kqueue`**